src/server/pps/server/api_server.go

```diff
@@ -3089,7 +3090,46 @@ func RepoNameToEnvString(repoName string) string {
 	return strings.ToUpper(repoName)
 }
 
-func (a *apiServer) rcPods(ctx context.Context, rcName string) ([]v1.Pod, error) {
+func (a *apiServer) pachdPods(ctx context.Context) ([]v1.Pod, error) {
+	podList, err := a.env.KubeClient.CoreV1().Pods(a.namespace).List(ctx, metav1.ListOptions{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ListOptions",
+			APIVersion: "v1",
+		},
+		LabelSelector: metav1.FormatLabelSelector(metav1.SetAsLabelSelector(map[string]string{"app": "pachd"})),
+	})
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	return podList.Items, nil
+}
+
+func (a *apiServer) rcPods(ctx context.Context, pipelineName string, pipelineVersion uint64) ([]v1.Pod, error) {
+	podList, err := a.env.KubeClient.CoreV1().Pods(a.namespace).List(ctx, metav1.ListOptions{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ListOptions",
+			APIVersion: "v1",
+		},
+		LabelSelector: metav1.FormatLabelSelector(metav1.SetAsLabelSelector(map[string]string{
+			"app":             "pipeline",
+			"pipelineName":    pipelineName,
+			"pipelineVersion": fmt.Sprint(pipelineVersion),
+		})),
+	})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return a.legacyRCPods(ctx, pipelineName, pipelineVersion)
+		}
+		return nil, errors.EnsureStack(err)
+	}
+	if len(podList.Items) == 0 {
+		return a.legacyRCPods(ctx, pipelineName, pipelineVersion)
+	}
+	return podList.Items, nil
+}
+
+func (a *apiServer) legacyRCPods(ctx context.Context, pipelineName string, pipelineVersion uint64) ([]v1.Pod, error) {
```

Comment:
I think this is worth a comment explaining in what sense it's legacy. I'm also not convinced we actually need this - this would be used when a new cluster is trying to iterate over pods from an out of date RC, but upgrading will trigger all of those to be deleted, which happens through the normal k8s propagation rather than directly via selector. Granted, this happens asynchronously, so it is possible to be in this state, so probably better safe than sorry.

Reply:
I added a comment in the equivalent location.